Use more smart pointers in markup.cpp

cdumez opened the pull request:
#### 96524b1516c8f6575ef510d38ca9d739f14a34a6
<pre>
Use more smart pointers in markup.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=261979">https://bugs.webkit.org/show_bug.cgi?id=261979</a>

Reviewed by Darin Adler.

* Source/WebCore/editing/ApplyStyleCommand.cpp:
(WebCore::ApplyStyleCommand::applyInlineStyleToNodeRange):
* Source/WebCore/editing/CompositeEditCommand.cpp:
(WebCore::CompositeEditCommand::moveParagraphContentsToNewBlockIfNecessary):
(WebCore::CompositeEditCommand::cleanupAfterDeletion):
(WebCore::CompositeEditCommand::moveParagraphs):
* Source/WebCore/editing/DeleteSelectionCommand.cpp:
(WebCore::isTableRow):
(WebCore::isTableCellEmpty):
(WebCore::isTableRowEmpty):
(WebCore::shouldRemoveContentOnly):
(WebCore::DeleteSelectionCommand::removePreviouslySelectedEmptyTableRows):
* Source/WebCore/editing/Editing.cpp:
(WebCore::isBlock):
(WebCore::isInline):
(WebCore::isTableStructureNode):
(WebCore::isListItem):
(WebCore::enclosingNodeOfType):
(WebCore::highestEnclosingNodeOfType):
(WebCore::enclosingListChild):
(WebCore::isTableCell):
(WebCore::numEnclosingMailBlockquotes):
(WebCore::isMailBlockquote):
* Source/WebCore/editing/Editing.h:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::wrappingStyleForSerialization):
* Source/WebCore/editing/FormatBlockCommand.cpp:
(WebCore::FormatBlockCommand::formatRange):
(WebCore::enclosingBlockToSplitTreeTo):
* Source/WebCore/editing/IndentOutdentCommand.cpp:
(WebCore::isListOrIndentBlockquote):
* Source/WebCore/editing/InsertListCommand.cpp:
(WebCore::InsertListCommand::listifyParagraph):
* Source/WebCore/editing/InsertParagraphSeparatorCommand.cpp:
(WebCore::isPhrasingContent):
(WebCore::InsertParagraphSeparatorCommand::doApply):
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::MarkupAccumulator):
(WebCore::MarkupAccumulator::startAppendingNode):
* Source/WebCore/editing/MarkupAccumulator.h:
* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
(WebCore::ReplacementFragment::removeUnrenderedNodes):
(WebCore::isMailPasteAsQuotationNode):
(WebCore::isHeaderElement):
(WebCore::ReplaceSelectionCommand::shouldMerge):
(WebCore::ReplaceSelectionCommand::removeRedundantStylesAndKeepStyleSpanInline):
(WebCore::ReplaceSelectionCommand::makeInsertedContentRoundTrippableWithHTMLTreeBuilder):
(WebCore::handleStyleSpansBeforeInsertion):
(WebCore::ReplaceSelectionCommand::handleStyleSpans):
(WebCore::isInlineNodeWithStyle):
(WebCore::nodeToSplitToAvoidPastingIntoInlineNodesWithStyle):
(WebCore::ReplaceSelectionCommand::doApply):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::shouldEmitTabBeforeNode):
(WebCore::shouldEmitNewlinesBeforeAndAfterNode):
* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::adjustSelectionRespectingGranularity):
* Source/WebCore/editing/VisibleUnits.cpp:
(WebCore::findStartOfParagraph):
(WebCore::findEndOfParagraph):
* Source/WebCore/editing/markup.cpp:
(WebCore::AttributeChange::AttributeChange):
(WebCore::completeURLs):
(WebCore::replaceSubresourceURLs):
(WebCore::removeSubresourceURLAttributes):
(WebCore::createPageForSanitizingWebContent):
(WebCore::sanitizeMarkup):
(WebCore::UserSelectNoneStateCache::computeState):
(WebCore::StyledMarkupAccumulator::StyledMarkupAccumulator):
(WebCore::StyledMarkupAccumulator::wrapWithNode):
(WebCore::StyledMarkupAccumulator::appendText):
(WebCore::StyledMarkupAccumulator::appendCustomAttributes):
(WebCore::StyledMarkupAccumulator::serializeNodes):
(WebCore::StyledMarkupAccumulator::traverseNodesForSerialization):
(WebCore::StyledMarkupAccumulator::appendNodeToPreserveMSOList):
(WebCore::needInterchangeNewlineAfter):
(WebCore::isElementPresentational):
(WebCore::highestAncestorToWrapMarkup):
(WebCore::serializePreservingVisualAppearanceInternal):
(WebCore::serializePreservingVisualAppearance):
(WebCore::serializeFragment):
(WebCore::fillContainerFromString):
(WebCore::isPlainTextMarkup):
(WebCore::contextPreservesNewline):
(WebCore::createFragmentFromText):
(WebCore::documentTypeString):
(WebCore::removeElementFromFragmentPreservingChildren):
(WebCore::canUseSetDataOptimization):
(WebCore::replaceChildrenWithFragment):
* Source/WebCore/editing/markup.h:
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
(WebCore::addSubresourcesForAttachmentElementsIfNecessary):
(WebCore::LegacyWebArchive::createFromSelection):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Source/WebCore/page/PageSerializer.cpp:
(WebCore::PageSerializer::SerializerMarkupAccumulator::SerializerMarkupAccumulator):
(WebCore::PageSerializer::serializeFrame):

Canonical link: <a href="https://commits.webkit.org/268378@main">https://commits.webkit.org/268378@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91f05bdfa7be6865c6c62d97e228a2d70aff3243

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19519 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20547 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21410 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18256 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19756 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23202 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20081 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19736 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19759 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16964 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22266 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16946 "7 flakes 4 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17752 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24063 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18007 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17927 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22035 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18536 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15705 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17680 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4668 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22036 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18364 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -104,8 +104,10 @@ class AttributeChange {
 
-    AttributeChange(Element* element, const QualifiedName& name, const AtomString& value)
-        : m_element(element), m_name(name), m_value(value)
+    AttributeChange(RefPtr<Element>&& element, const QualifiedName& name, const AtomString& value)
```
We should move in the value instead of copying. Could change the type of the argument to `AtomString`. In fact, since this stores the name and value neither should be `const&`; never does any good.

darinadler commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -838,23 +835,23 @@ Node* StyledMarkupAccumulator::traverseNodesForSerialization(Node& startNode, No
     if (is<Comment>(node)) {
-        auto& commentNode = downcast<Comment>(node);
-        if (!m_inMSOList && commentNode.data() == "[if !supportLists]"_s)
+        Ref commentNode = downcast<Comment>(node);
```
Surprised you didn’t switch to `dynamicDowncast`.

darinadler commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -838,23 +835,23 @@ Node* StyledMarkupAccumulator::traverseNodesForSerialization(Node& startNode, No
 
-        auto& textChild = downcast<Text>(*firstChild);
-        auto& styleContent = textChild.data();
+        auto textChild = downcast<Text>(firstChild.releaseNonNull());
```
Surprised you didn’t switch to `dynamicDowncast`.

cdumez committed:
Use more smart pointers in markup.cpp
https://bugs.webkit.org/show_bug.cgi?id=261979

Reviewed by Darin Adler.

* Source/WebCore/editing/ApplyStyleCommand.cpp:
(WebCore::ApplyStyleCommand::applyInlineStyleToNodeRange):
* Source/WebCore/editing/CompositeEditCommand.cpp:
(WebCore::CompositeEditCommand::moveParagraphContentsToNewBlockIfNecessary):
(WebCore::CompositeEditCommand::cleanupAfterDeletion):
(WebCore::CompositeEditCommand::moveParagraphs):
* Source/WebCore/editing/DeleteSelectionCommand.cpp:
(WebCore::isTableRow):
(WebCore::isTableCellEmpty):
(WebCore::isTableRowEmpty):
(WebCore::shouldRemoveContentOnly):
(WebCore::DeleteSelectionCommand::removePreviouslySelectedEmptyTableRows):
* Source/WebCore/editing/Editing.cpp:
(WebCore::isBlock):
(WebCore::isInline):
(WebCore::isTableStructureNode):
(WebCore::isListItem):
(WebCore::enclosingNodeOfType):
(WebCore::highestEnclosingNodeOfType):
(WebCore::enclosingListChild):
(WebCore::isTableCell):
(WebCore::numEnclosingMailBlockquotes):
(WebCore::isMailBlockquote):
* Source/WebCore/editing/Editing.h:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::wrappingStyleForSerialization):
* Source/WebCore/editing/FormatBlockCommand.cpp:
(WebCore::FormatBlockCommand::formatRange):
(WebCore::enclosingBlockToSplitTreeTo):
* Source/WebCore/editing/IndentOutdentCommand.cpp:
(WebCore::isListOrIndentBlockquote):
* Source/WebCore/editing/InsertListCommand.cpp:
(WebCore::InsertListCommand::listifyParagraph):
* Source/WebCore/editing/InsertParagraphSeparatorCommand.cpp:
(WebCore::isPhrasingContent):
(WebCore::InsertParagraphSeparatorCommand::doApply):
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::MarkupAccumulator):
(WebCore::MarkupAccumulator::startAppendingNode):
* Source/WebCore/editing/MarkupAccumulator.h:
* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
(WebCore::ReplacementFragment::removeUnrenderedNodes):
(WebCore::isMailPasteAsQuotationNode):
(WebCore::isHeaderElement):
(WebCore::ReplaceSelectionCommand::shouldMerge):
(WebCore::ReplaceSelectionCommand::removeRedundantStylesAndKeepStyleSpanInline):
(WebCore::ReplaceSelectionCommand::makeInsertedContentRoundTrippableWithHTMLTreeBuilder):
(WebCore::handleStyleSpansBeforeInsertion):
(WebCore::ReplaceSelectionCommand::handleStyleSpans):
(WebCore::isInlineNodeWithStyle):
(WebCore::nodeToSplitToAvoidPastingIntoInlineNodesWithStyle):
(WebCore::ReplaceSelectionCommand::doApply):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::shouldEmitTabBeforeNode):
(WebCore::shouldEmitNewlinesBeforeAndAfterNode):
* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::adjustSelectionRespectingGranularity):
* Source/WebCore/editing/VisibleUnits.cpp:
(WebCore::findStartOfParagraph):
(WebCore::findEndOfParagraph):
* Source/WebCore/editing/markup.cpp:
(WebCore::AttributeChange::AttributeChange):
(WebCore::completeURLs):
(WebCore::replaceSubresourceURLs):
(WebCore::removeSubresourceURLAttributes):
(WebCore::createPageForSanitizingWebContent):
(WebCore::sanitizeMarkup):
(WebCore::UserSelectNoneStateCache::computeState):
(WebCore::StyledMarkupAccumulator::StyledMarkupAccumulator):
(WebCore::StyledMarkupAccumulator::wrapWithNode):
(WebCore::StyledMarkupAccumulator::appendText):
(WebCore::StyledMarkupAccumulator::appendCustomAttributes):
(WebCore::StyledMarkupAccumulator::serializeNodes):
(WebCore::StyledMarkupAccumulator::traverseNodesForSerialization):
(WebCore::StyledMarkupAccumulator::appendNodeToPreserveMSOList):
(WebCore::needInterchangeNewlineAfter):
(WebCore::isElementPresentational):
(WebCore::highestAncestorToWrapMarkup):
(WebCore::serializePreservingVisualAppearanceInternal):
(WebCore::serializePreservingVisualAppearance):
(WebCore::serializeFragment):
(WebCore::fillContainerFromString):
(WebCore::isPlainTextMarkup):
(WebCore::contextPreservesNewline):
(WebCore::createFragmentFromText):
(WebCore::documentTypeString):
(WebCore::removeElementFromFragmentPreservingChildren):
(WebCore::canUseSetDataOptimization):
(WebCore::replaceChildrenWithFragment):
* Source/WebCore/editing/markup.h:
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
(WebCore::addSubresourcesForAttachmentElementsIfNecessary):
(WebCore::LegacyWebArchive::createFromSelection):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Source/WebCore/page/PageSerializer.cpp:
(WebCore::PageSerializer::SerializerMarkupAccumulator::SerializerMarkupAccumulator):
(WebCore::PageSerializer::serializeFrame):

Canonical link: https://commits.webkit.org/268378@main
